include/deal.II/non_matching/fe_values.h

```diff
@@ -0,0 +1,511 @@
+     */
+    LocationToLevelSet current_cell_location;
+
+    /**
+     * The update flags passed to the constructor.
+     */
+    const RegionUpdateFlags region_update_flags;
+
+    /**
+     * Pointer to the MeshClassifier passed to the constructor.
+     */
+    const SmartPointer<const MeshClassifier<dim>> mesh_classifier;
+
+    /**
+     * hp::FEValues object created with the UpdateFlags for the inside region
+     * and quadrature rules over the full reference cell: $[0, 1]^dim$, i.e.,
+     * the QCollection passed to the constructor.
+     *
+     * This object is used to get dealii::FEValues objects when the cell is not
+     * intersected and we don't need to generate immersed quadrature rules.
+     */
+    hp::FEValues<dim> fe_values_inside_full_quadrature;
+
+    /**
+     * hp::FEValues object created with the UpdateFlags for the outside region
+     * and quadrature rules over the full reference cell: $[0, 1]^dim$, i.e.,
+     * the QCollection passed to the constructor.
+     *
+     * This object is used to get dealii::FEValues objects when the cell is not
+     * intersected and we don't need to generate immersed quadrature rules.
+     */
+    hp::FEValues<dim> fe_values_outside_full_quadrature;
+
+    /**
+     * Pointer to an FEValues object created with a quadrature integrating over
+     * the inside region, $\{x \in B : \psi(x) < 0 \}$, that was generated in
+     * the last call to reinit(..). If the cell in the last call was not
+     * intersected or if 0 quadrature points were generated, the unique_ptr will
+     * we empty.
+     */
+    std::unique_ptr<dealii::FEValues<dim>> fe_values_inside;
+
+    /**
+     * Pointer to an FEValues object created with a quadrature integrating over
+     * the inside region, $\{x \in B : \psi(x) > 0 \}$, that was generated in
+     * the last call to reinit(..). If the cell in the last call was not
+     * intersected or if 0 quadrature points were generated, the unique_ptr will
+     * we empty.
+     */
+    std::unique_ptr<dealii::FEValues<dim>> fe_values_outside;
+
+    /**
+     * Pointer to an FEImmersedSurfaceValues object created with a quadrature
+     * integrating over the surface region, $\{x \in B : \psi(x) = 0 \}$, that
+     * was generated in the last call to reinit(..). If the cell in the last
+     * call was not intersected or if 0 quadrature points were generated, the
+     * unique_ptr will we empty.
+     */
+    std::unique_ptr<NonMatching::FEImmersedSurfaceValues<dim>>
+      fe_values_surface;
+
+    /**
+     * Object that generates the immersed quadrature rules.
+     */
+    QuadratureGenerator<dim> quadrature_generator;
+
+    /**
+     * Function that describes our level set function in reference space.
+     */
+    const std::unique_ptr<
+      internal::FEValuesImplementation::CellWiseFunction<dim>>
+      ref_space_level_set;
+  };
+
+
+  namespace internal
+  {
+    namespace FEValuesImplementation
+    {
```

Comment:
I think everything below here can go into the source file.

Reply:
I had a test for `RefSpaceFEFieldFunction`, but I had forgotten to add it. That is why this is in the header. We can either keep this in the header or remove the test. What do you think is best?

Reply:
I would remove the test and move the content to the source file. `RefSpaceFEFieldFunction` indirectly via `NonMatching::FEValues`.